pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -0,0 +1,181 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package trustbundle
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"sync"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/pkg/log"
+)
+
+// source is all possible sources of MeshConfig
+type Source int
+
+type TrustAnchorConfig struct {
+	Source Source
+	Certs  []string
+}
+
+type TrustAnchorUpdate struct {
+	TrustAnchorConfig
+}
+
+type TrustBundle struct {
+	sourceConfig map[Source]TrustAnchorConfig
+	mutex        sync.RWMutex
+	mergedCerts  []string
+	updatecb     func()
+}
+
+var (
+	trustBundleLog = log.RegisterScope("trustBundle", "Workload mTLS trust bundle logs", 0)
+
+	// GlobalWorkloadTrustbundle Needed to interface with XDS
+	GlobalWorkloadTrustbundle *TrustBundle = nil
+)
+
+const (
+	SourceIstioCA Source = iota
+	SourceMeshConfig
+	SourceIstioRA
+	// SOURCE_SPIFFE_ENDPOINT
+	UpdateChannelLen = 10
+)
+
+func checkSameCerts(certs1 []string, certs2 []string) bool {
+	if len(certs1) != len(certs2) {
+		return false
+	}
+	for i := range certs1 {
+		if certs1[i] != certs2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func NewTrustBundle(updatecb func()) *TrustBundle {
+	tb := &TrustBundle{
+		sourceConfig: map[Source]TrustAnchorConfig{
+			SourceIstioCA:    {Source: SourceIstioCA, Certs: []string{}},
+			SourceMeshConfig: {Source: SourceMeshConfig, Certs: []string{}},
+			SourceIstioRA:    {Source: SourceIstioRA, Certs: []string{}},
+			// SOURCE_SPIFFE_ENDPOINT:        &TrustAnchorConfig{source: SOURCE_SPIFFE_ENDPOINT, certs: []string{}},
+		},
+		mergedCerts: []string{},
+		updatecb:    updatecb,
+	}
+
+	GlobalWorkloadTrustbundle = tb
+	return tb
+}
+
+// GetTrustBundle : Retrieves all the trustAnchors for current Spiffee Trust Domain
+func (tb *TrustBundle) GetTrustBundle() []string {
+	tb.mutex.RLock()
+	defer tb.mutex.RUnlock()
+	trustedCerts := make([]string, len(tb.mergedCerts))
+	copy(trustedCerts, tb.mergedCerts)
+	return trustedCerts
+}
+
+func (tb *TrustBundle) verifyTrustAnchor(trustAnchor string) error {
+	block, _ := pem.Decode([]byte(trustAnchor))
+	if block == nil {
+		return fmt.Errorf("failed to decode pem certificate")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse X.509 certificate")
+	}
+	if !cert.IsCA {
+		return fmt.Errorf("certificate is not a CA certificate")
+	}
+	return nil
+}
+
+func (tb *TrustBundle) mergeInternal() {
+	var ok bool
+	mergeCerts := []string{}
+	certMap := make(map[string]bool)
+
+	tb.mutex.Lock()
+	defer tb.mutex.Unlock()
+
+	for _, configSource := range tb.sourceConfig {
+		for _, cert := range configSource.Certs {
+			if _, ok = certMap[cert]; !ok {
+				certMap[cert] = true
+				mergeCerts = append(mergeCerts, cert)
+			}
+		}
+	}
+	tb.mergedCerts = mergeCerts
+}
+
+// External Function to merge a TrustAnchor config with the existing TrustBundle
```

Comment:
nit: go best practice is to start comment with the function name like `// UpdateTrustAnchor blah blah`.

We don't need to document its external as well, as go automatically does this for us by the function name